lib/puppet_x/relay/agent/job/exec.rb

```diff
@@ -0,0 +1,94 @@
+require 'puppet'
+require 'puppet/util/pidlock'
+
+require_relative '../backend'
+require_relative '../model'
+require_relative 'base'
+require_relative 'schedule'
+
+module PuppetX
+  module Relay
+    module Agent
+      module Job
+        class Exec < Base
+          # @param backend [Backend::Base]
+          # @param run [Model::Run]
+          # @param state_dir [String]
+          def initialize(backend, run, state_dir)
+            @backend = backend
+            @run = run
+            @state_dir = state_dir
+            @accepted = false
+            @retries = 3
+          end
+
+          def handle(job)
+            Puppet::FileSystem.mkpath(@state_dir)
+
+            @state_lock ||= Puppet::Util::Pidlock.new(File.join(@state_dir, 'state'))
+            unless @state_lock.lock
+              # Running concurrently with another process.
+              job.disable
```

Comment:
Would this mean that if two runs on the same node were requested, it could possible run a single time?

Reply:
This state directory is already scoped to the run ID from the Relay API, so it should only prevent race conditions within the same run, not across different runs.